### PR TITLE
Closes #22: Use vocabulary management

### LINKED ISSUE #22: Use vocabulary management interface instead of directly transacting schema datoms
Here's how that looks (docs + test):

https://github.com/mozilla/mentat/blob/master/src/vocabulary.rs#L12

### PR DESCRIPTION
**Pull request by fluffyemily**
Contemplating getting rid of the store crate altogether, but I will wait until I see what it looks like after I've done #21 and then decide.

I actually think this makes it harder to see at a glance what the schema looks like, but it's easier to construct. Maybe I'll add a comment in datalog....

### CODE REVIEW

File: rust/src/lib.rs [hunk 1]
```diff
@@ -207,15 +225,17 @@ impl Toodle {
                         [?l :label/name ?name]
                         [?l :label/color ?color]
         ]"#;
-        self.connection
-            .query_args(query, vec![(Variable::from_valid_name("?item_uuid"), item_uuid.to_typed_value())])
+        let in_progress_read = self.connection.begin_read().expect("expected a read transaction");
```
Comment (rnewman) [inline rust/src/lib.rs]: Don't use `expect` for these; that's just `unwrap` with a nice message! You're in a function that returns `Result`, so just `?`. If that doesn't work, you need to add a `From` implementation somewhere.

File: rust/src/lib.rs [hunk 2]
```diff
@@ -224,65 +244,84 @@ impl Toodle {
...
                         [?eid :item/name ?name]
         ]"#;
-        self.connection
-            .query_args(query, vec![(Variable::from_valid_name("?label"), label.name.to_typed_value())])
-            .into_rel_result()
-            .map(|rows| rows.into_iter().map(|r| self.item_row_to_item(r)).collect())
+        let rows;
+        {
+            let in_progress_read = self.connection.begin_read()?;
+            let args = QueryInputs::with_value_sequence(vec![(Variable::from_valid_name("?label"), label.name.to_typed_value())]);
```
Comment (rnewman) [inline rust/src/lib.rs]: Do we need a `var!` macro?

Comment (rnewman) [reply rust/src/lib.rs]: Filed https://github.com/mozilla/mentat/issues/548

File: rust/src/lib.rs [hunk 3]
```diff
@@ -224,65 +244,84 @@ impl Toodle {
...
+            rows = in_progress_read
+                .q_once(query, args)
+                .into_rel_result();
+        }
+        rows.map(|rows| rows.into_iter().map(|r| self.item_row_to_item(r)).collect())
             .map_err(|e| e.into())
```
Comment (rnewman) [inline rust/src/lib.rs]: We should be able to replace these `map_err` lines with `?`.

File: rust/src/lib.rs [hunk 4]
```diff
@@ -79,45 +92,65 @@ use store::{
...
+                (kw!(:item/label),
+                 AttributeBuilder::new()
+                    .value_type(ValueType::Ref)
+                    .multival(true)
+                    .build()),
+            ],
+        })
         .map_err(|e| e.into())
-        .map(|_| ())
+        .map(|_|())
```
Comment (rnewman) [inline rust/src/lib.rs]: This is just `.and(Ok(()))`. Or you can do this:

```
impl<T, E> Result<T, E> {
    fn void(self) -> Result<(), E> {
        self.and(Ok(()))
    }
}
```

and do

```
.void()
```

here.

File: rust/src/lib.rs [hunk 5]
```diff
@@ -79,45 +92,65 @@ use store::{
...
+                (kw!(:label/color),
+                 AttributeBuilder::new()
+                    .value_type(ValueType::String)
+                    .multival(false)
+                    .build()),
+            ],
+        })
         .map_err(|e| e.into())
-        .map(|_| ())
+        .map(|_|())
```
Comment (rnewman) [inline rust/src/lib.rs]: If you do it a lot, `.void()` might be worthwhile.

File: rust/src/lib.rs [hunk 6]
```diff
@@ -79,45 +92,65 @@ use store::{
...
-            :db/valueType   :db.type/instant
-            :db/cardinality :db.cardinality/one  },
-        {   :db/ident       :item/completion_date
-            :db/valueType   :db.type/instant
-            :db/cardinality :db.cardinality/one  },
-        {  :db/ident     :item/label
-            :db/valueType :db.type/ref
-            :db/cardinality :db.cardinality/many }]"#;
-    in_progress.transact(schema)
+    in_progress.ensure_vocabulary(&Definition {
```
Comment (rnewman) [inline rust/src/lib.rs]: Informal before/after timings and memory usage of this (in release mode)? I would expect a small memory saving…

**Review by rnewman — APPROVED** [review]

File: rust/src/lib.rs (hunk 1 above)
Comment (fluffyemily) [reply rust/src/lib.rs]: All the others were `?` so I think I just missed this one.

File: rust/src/lib.rs (hunk 6 above)
Comment (fluffyemily) [reply rust/src/lib.rs]: Marginally faster startup times: 
Non vocab management:
```
startup time: Duration { secs: 0, nanos: 783583278 }
startup time: Duration { secs: 0, nanos: 718939582 }
startup time: Duration { secs: 0, nanos: 428831956 }
startup time: Duration { secs: 0, nanos: 61791586 }
startup time: Duration { secs: 0, nanos: 63743039 }
```

Vocab management:
```
startup time: Duration { secs: 0, nanos: 769641038 }
startup time: Duration { secs: 0, nanos: 679142944 }
startup time: Duration { secs: 0, nanos: 330097111 }
startup time: Duration { secs: 0, nanos: 58794550 }
startup time: Duration { secs: 0, nanos: 54428597 }
```

**Pull request merged**